### PR TITLE
Encoding of response with empty body

### PR DESCRIPTION
**Pull request by christiankjaer**
Fixes https://github.com/http4s/http4s/issues/6427

Around the repository there are different ways that empty responses are encoded. In some places the content length is set to 0, and in other places like the auth middleware, an empty response is returned with no headers set.

Here is a potential fix where an empty body with `content-length: 0` is always set when the status code allows a non-empty response and the header is not already set.

Something similar could probably also be done in the request encoding as well.

### CODE REVIEW

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala [hunk 1]
```diff
@@ -63,8 +64,12 @@ private[ember] object Encoder {
           ()
         }
       }
-      if (!chunked && !appliedContentLength && resp.status.isEntityAllowed) {
-        stringBuilder.append(chunkedTansferEncodingHeaderRaw).append(CRLF)
+      if (!appliedContentLength && resp.body == EmptyBody && resp.status.isEntityAllowed) {
```
Comment (rossabaker) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:67]: The logic here looks correct, but it's observable behavior that's dependent on reference equality of a stream.  They're both "correct" HTTP, and this is a better outcome for a common case.  I think I'm okay with this, but calling it out.

In main, with the new entity model, we can target the Empty entity and not have this concern.

**Review by rossabaker — APPROVED** [review]
:+1:, but I'd like to hear someone else's thought on my worry below.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala (hunk 1 above)
Comment (christiankjaer) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:67]: I looked at it, and I could not see any nice way to check whether a fs2 stream is empty. Maybe somebody can enlighten me.

Comment (christiankjaer) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:67]: I am still confused why it was broken with an empty body and the chunked encoding.

Comment (rossabaker) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:67]: There really isn't a good way.  We can "peek" a stream, but that semantically blocks until the first chunk of the body (if any) is produced, and would delay rendering the headers.  In scalaz-stream, we used to be able to introspect the algebra and find a wider class of empty streams, but still not all off them.  Effects are, well, effectful.

Comment (christiankjaer) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:67]: Ah ok. Makes sense :+1:

File: ember-core/shared/src/test/scala/org/http4s/ember/core/EncoderSuite.scala [hunk 2]
```diff
@@ -125,6 +126,32 @@ class EncoderSuite extends Http4sSuite {
...
+    val expected =
+      """HTTP/1.1 200 OK
+      |Transfer-Encoding: chunked
+      |
+      |0
+      |
+      |""".stripMargin
+
+    Helpers.encodeResponseRig(resp).assertEquals(expected)
+  }
```
Comment (christiankjaer) [inline ember-core/shared/src/test/scala/org/http4s/ember/core/EncoderSuite.scala:153]: So is this intended behavior? I guess it's the only implementation that makes sense.

Comment (rossabaker) [reply ember-core/shared/src/test/scala/org/http4s/ember/core/EncoderSuite.scala:153]: Yeah, I don't think we can do better than this.  And even though it's arguably overspecified, I like the test to alert us to behavior changes.

**Comment by rossabaker** [comment]
This will need an import sort.  I think `ember-core/scalafix --all` will do it.

**Review by rossabaker — APPROVED** [review]
:+1: if green

**Comment by christiankjaer** [comment]
Only question is whether to actually have `main` as the base branch, since I observe the bug only on the 1.0 pre release.

**Comment by rossabaker** [comment]
Do those tests pass on this branch without your change?  I think only one of them would?

We will want to tweak this on the merge to main, where we can detect empty entities (it's a sum type, so no FS2 introspection is involved).

**Comment by christiankjaer** [comment]
Yes. Only one will pass.

**Comment by rossabaker** [comment]
I'd say this is correctly targeted for 0.23 then.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala (hunk 1 above)
Comment (armanbilge) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:67]: > I think I'm okay with this, but calling it out.

Seems fine to me.

**Review by armanbilge — APPROVED** [review]
Nice work!

**Pull request merged**